### PR TITLE
[capi-media-vision][ACR-1700] Add face recognition API guide

### PR DESCRIPTION
**Pull request by daeinki**
Signed-off-by: Inki Dae <inki.dae@samsung.com>

Add deep learning based face recognition API guide for Mediavision.
ACR: ACR-1700

### CODE REVIEW

**Comment by safir-srbd** [comment]
@daeinki , hello. Hope you are well!

Thank you for pushing the changes to the appropriate branch. 
However, with a quick glance, I can see that you haven't added the newly created file to the `toc_all.md` file. Please add the file to the `toc_all.md` file so that it appears properly in the left navigation bar.

**Comment by mijin85cho** [comment]
> @daeinki , hello. Hope you are well!
> 
> Thank you for pushing the changes to the appropriate branch. However, with a quick glance, I can see that you haven't added the newly created file to the `toc_all.md` file. Please add the file to the `toc_all.md` file so that it appears properly in the left navigation bar.

Also, please don't forget to check the overview file. In this case application/native/guides/multimedia/overview.md should be updated to add the sub page.

File: docs/application/native/guides/multimedia/face-recognition.md [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+   3.3 Recognize the given face image, which is a process that face recognition framework performs a forward propergation to find the most suitable label.
+
+   ```
+   // Recognize the given face image.
+   ret = mv_face_recognition_inference(handle, mv_source);
+   if (ret != MEDIA_VISION_ERROR_NONE) {
+       // handle an error.
+   }
+   ```
```
Comment (safir-srbd) [inline docs/application/native/guides/multimedia/face-recognition.md:145]:
```suggestion
    3.  Recognize the given face image, which is a process in which the face recognition framework performs a forward propagation to find the most suitable label.

        ```
        // Recognize the given face image.
        ret = mv_face_recognition_inference(handle, mv_source);
        if (ret != MEDIA_VISION_ERROR_NONE) {
            // handle an error.
        }
        ```
```

File: docs/application/native/guides/multimedia/face-recognition.md [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+4. Get the label string as a result. This is a process that face recognition framework returns a inference result as a label string
+   which was used when training. Please keep in mind. This function is valid only after mv_face_recognition_inference function call is completed.
```
Comment (safir-srbd) [inline docs/application/native/guides/multimedia/face-recognition.md:148]:
```suggestion
4. Get the label string as a result. This is a process in which the face recognition framework returns an inference result as a label string which was used when training. Please keep in mind that the function is valid only after mv_face_recognition_inference function call is completed. This is shown in the code snippet below:
```

**Review by safir-srbd — COMMENTED** [review]
@daeinki I have reviewed this PR, and have fixed some formatting and grammar issues. Please incorporate these suggestions alongside the comments that have been given above.

**Comment by daeinki** [comment]
> @daeinki , hello. Hope you are well!
> 
> Thank you for pushing the changes to the appropriate branch. However, with a quick glance, I can see that you haven't added the newly created file to the `toc_all.md` file. Please add the file to the `toc_all.md` file so that it appears properly in the left navigation bar.

Got it. Thanks.

File: docs/application/native/guides/multimedia/face-recognition.md [hunk 3]
```diff
@@ -0,0 +1,172 @@
+# Deep learning based Face Recognition
```
[the remark this answers is not in the document]
Comment (daeinki) [reply docs/application/native/guides/multimedia/face-recognition.md:1]: > Title Case is recommended for First Headings

Thanks.

File: docs/application/native/guides/multimedia/face-recognition.md [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+You can perceive and understand a face in your application.
+This document guides how to register a new face, to unregister someone who is registered
+with a given label and to recognize a given face.
```
[the remark this answers is not in the document]
Comment (daeinki) [reply docs/application/native/guides/multimedia/face-recognition.md:5]: > Using deep learning based face recognition, you can perceive and understand faces within your application.

Thanks.

File: docs/application/native/guides/multimedia/face-recognition.md [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+## Prerequisites
+
+To enable your application for using the deep learning based face recognition functionality:
+
+1. To use the functions and data types of the deep learning based face recognition API, include the `<mv_face_recognition.h>` header file in your application.
+
+   In addition, you may need to include the `<image_util.h>` header file to handle the image decoding task, or the `<camera.h>` header file to provide preview images.
+   Of course, you could use other libraries you want.
```
[the remark this answers is not in the document]
Comment (daeinki) [reply docs/application/native/guides/multimedia/face-recognition.md:14]: Thanks.

File: docs/application/native/guides/multimedia/face-recognition.md [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+   ```
+   #include <mv_face_recognition.h>
+
+   /* To decode the image file. */
+   #include <image_util.h>
+   /* To get preview image from Camera device. */
+   #include <camera.h>
+   ```
+
```
[the remark this answers is not in the document]
Comment (daeinki) [reply docs/application/native/guides/multimedia/face-recognition.md:24]: Thanks.

File: docs/application/native/guides/multimedia/face-recognition.md [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+To register a new face image with a given label:
```
[the remark this answers is not in the document]
Comment (daeinki) [reply docs/application/native/guides/multimedia/face-recognition.md:25]: Thanks.

File: docs/application/native/guides/multimedia/face-recognition.md [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+
+1. Create a face recognition handle and prepare for the face recognition framework.
```
[the remark this answers is not in the document]
Comment (daeinki) [reply docs/application/native/guides/multimedia/face-recognition.md:27]: Thanks.

File: docs/application/native/guides/multimedia/face-recognition.md [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+   ret = mv_face_recognition_prepare(handle);
+   if (ret != MEDIA_VISION_ERROR_NONE) {
+       // handle an error.
+   }
+   ```
+
+2. Prepare an input source to register. The input source should be RGB data. In this guide,
+   we will use a image file as a input source, and we will show you how to decode the given image file
+   and then prepare for a mv_source(Mediavison source) which contains the decoded image buffer.
+   All Mediavision API needs the Mediavision source handle as a input, and that is mandatorily required.
```
[the remark this answers is not in the document]
Comment (daeinki) [reply docs/application/native/guides/multimedia/face-recognition.md:47]: Thanks.

File: docs/application/native/guides/multimedia/face-recognition.md [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+   ret = image_util_decode_destroy(imageDecoder);
+   if (ret != IMAGE_UTIL_ERROR_NONE) {
+       // handle an error.
+   }
+
+   free(dataBuffer);
+   ```
+
+3. Now, we are ready for using the face recognition main features. There are three face recognition features - register, unregister and recognize.
+   You can use one of them.
```
[the remark this answers is not in the document]
Comment (daeinki) [reply docs/application/native/guides/multimedia/face-recognition.md:106]: Thanks.

File: docs/application/native/guides/multimedia/face-recognition.md [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+ 
+   3.1 Register a new face, which is a process that the face recognition framework tains and generates
+       a internal model file with the given face data and its label. You could repeat this step to train
+	   the internal model file with more faces. For the accuracy, we recommend to train at least three images per a face.
```
[the remark this answers is not in the document]
Comment (daeinki) [reply docs/application/native/guides/multimedia/face-recognition.md:110]: Thanks.

File: docs/application/native/guides/multimedia/face-recognition.md [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+   ret = mv_face_recognition_register(handle, mv_source, "name_a");
+   if (ret != MEDIA_VISION_ERROR_NONE) {
+      // handle an error.
+   }
+
+   ret = mv_destroy_source(mv_source);
+   if (ret != MEDIA_VISION_ERROR_NONE) {
+       // handle an error.
+   }
+   ```
```
[the remark this answers is not in the document]
Comment (daeinki) [reply docs/application/native/guides/multimedia/face-recognition.md:123]: Thanks.

File: docs/application/native/guides/multimedia/face-recognition.md [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+       recognition framework drops all data related to the given label string. In fact, this process is to retrain the internal model
+       without the dropped data.
+
+   ```
+   // Drop all data related to a given label string - "name_a".
+   ret = mv_face_recognition_unregister(handle, "name_a");
+   if (ret != MEDIA_VISION_ERROR_NONE) {
+       // handle an error.
+   }
+   ```
```
[the remark this answers is not in the document]
Comment (daeinki) [reply docs/application/native/guides/multimedia/face-recognition.md:135]: Thanks.

File: docs/application/native/guides/multimedia/face-recognition.md (hunk 1 above)
Comment (daeinki) [reply docs/application/native/guides/multimedia/face-recognition.md:145]: Thanks.

File: docs/application/native/guides/multimedia/face-recognition.md (hunk 2 above)
Comment (daeinki) [reply docs/application/native/guides/multimedia/face-recognition.md:148]: Thanks.

File: docs/application/native/guides/multimedia/face-recognition.md [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+   // Get a label string as a face recognition result.
+   const char *out_label = NULL;
+   ret = mv_face_recognition_get_label(handle, &out_label);
+   if (ret != MEDIA_VISION_ERROR_NONE) {
+       // handle an error.
+   }
+   ```
+
+5. Destroy the face recognition handle. This is a process to release all resources used. Please keep in mind, the out_label variable at step 4 is valid
+   only while the face recognition handle is alive.
```
[the remark this answers is not in the document]
Comment (daeinki) [reply docs/application/native/guides/multimedia/face-recognition.md:160]: Thanks.

File: docs/application/native/guides/multimedia/face-recognition.md [hunk 15]
```diff
@@ -0,0 +1,157 @@
+# Deep Learning Based Face Recognition
+
+Using deep learning based face recognition, you can perceive and understand faces within your application.
+This document is a guide on how to register a new face, how to unregister a face that is registered with a given label, and how to recognize a given face.
```
[the remark this answers is not in the document]
Comment (daeinki) [reply docs/application/native/guides/multimedia/face-recognition.md:4]: Thanks.

File: docs/application/native/guides/multimedia/face-recognition.md [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+    Keep in mind that you could use other libraries if required.
+
+   ```
+   #include <mv_face_recognition.h>
+
+   /* To decode the image file. */
+   #include <image_util.h>
+   /* To get preview image from Camera device. */
+   #include <camera.h>
+   ```
```
[the remark this answers is not in the document]
Comment (daeinki) [reply docs/application/native/guides/multimedia/face-recognition.md:20]: Thanks.

File: docs/application/native/guides/multimedia/face-recognition.md [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+   const char *out_label = NULL;
+   ret = mv_face_recognition_get_label(handle, &out_label);
+   if (ret != MEDIA_VISION_ERROR_NONE) {
+       // handle an error.
+   }
+   ```
+
+5. Destroy the face recognition handle. This is a process to release all resources used. Please keep in mind, the out_label variable at step 4 is valid only while the face recognition handle is alive.
+
+```
```
[the remark this answers is not in the document]
Comment (daeinki) [reply docs/application/native/guides/multimedia/face-recognition.md:147]: Thanks.

File: docs/application/native/guides/multimedia/face-recognition.md [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+   ret = mv_face_recognition_destroy(handle);
+   if (ret != MEDIA_VISION_ERROR_NONE) {
+       // handle an error.
```
[the remark this answers is not in the document]
Comment (daeinki) [reply docs/application/native/guides/multimedia/face-recognition.md:150]: Thanks.

File: docs/application/native/guides/multimedia/face-recognition.md [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+   }
+```
```
[the remark this answers is not in the document]
Comment (daeinki) [reply docs/application/native/guides/multimedia/face-recognition.md:152]: Thanks.

File: docs/application/native/guides/multimedia/media-vision.md [hunk 20]
```diff
@@ -8,6 +8,9 @@ You can use the following visual detection and recognition features in your nati
   
   You can perceive and understand an image or extract information from images in your application. You can encrypt a message by generating a barcode from it, and you can also detect barcodes in an image or from a camera preview stream.
 - [Face Detection, Recognition, and Tracking](face-detection.md)
+
+  You can register, unregister and recognize a face in your application with more accuracy, which uses Deep Learning technique.
+- [Deep Learning Based Face Recognition](face-recognition.md)
   
   You can detect a face from an image, and recognize a face with a set of examples. You can also track a face within the camera preview stream.
```
[the remark this answers is not in the document]
Comment (daeinki) [reply docs/application/native/guides/multimedia/media-vision.md:15]: Thanks.

File: docs/application/native/guides/multimedia/face-recognition.md [hunk 21]
```diff
@@ -0,0 +1,156 @@
...
+5. Destroy the face recognition handle. This is a process to release all resources used. Please keep in mind, the out_label variable at step 4 is valid only while the face recognition handle is alive.
+
+    ```
+    ret = mv_face_recognition_destroy(handle);
+    if (ret != MEDIA_VISION_ERROR_NONE) {
+        // handle an error.
+    }
+    ```
+
+## Related Information
```
Comment (safir-srbd) [inline docs/application/native/guides/multimedia/face-recognition.md:153]: Second headings must follow sentence case.
Ref: https://github.com/Samsung/tizen-docs/blob/master/styleguide/style.md#capitalization
```suggestion
## Related information
```

**Review by safir-srbd — COMMENTED** [review]
@daeinki Thank you so much for addressing the issues in such quick time. After giving another check, I found one small issue. Could you please address that?

File: docs/application/native/guides/multimedia/media-vision.md [hunk 22]
```diff
@@ -7,15 +7,23 @@ You can use the following visual detection and recognition features in your nati
...
   You can perceive and understand an image or extract information from images in your application. You can encrypt a message by generating a barcode from it, and you can also detect barcodes in an image or from a camera preview stream.
+
 - [Face Detection, Recognition, and Tracking](face-detection.md)
-  
+
   You can detect a face from an image, and recognize a face with a set of examples. You can also track a face within the camera preview stream.
+
+- [Deep Learning Based Face Recognition](face-recognition.md)
+
+  You can register, unregister and recognize a face in your application with more accuracy, which uses Deep Learning technique.
```
Comment (safir-srbd) [inline docs/application/native/guides/multimedia/media-vision.md:17]: @daeinki Also I think these headings should be placed in alphabetical order. So Deep Learning Based Face Recognition should be placed before Face Detection, Recognition, and Tracking. Please address this

File: docs/application/native/guides/multimedia/face-recognition.md (hunk 21 above)
Comment (daeinki) [reply docs/application/native/guides/multimedia/face-recognition.md:153]: Thanks.

File: docs/application/native/guides/multimedia/media-vision.md (hunk 22 above)
Comment (daeinki) [reply docs/application/native/guides/multimedia/media-vision.md:17]: Got it. Thanks.

**Review by safir-srbd — COMMENTED** [review]
@daeinki Thank you so much for implementing the suggested changes.

@asm-foysal @mijin85cho I am approving this PR for merging.

**Review by safir-srbd — APPROVED** [review]

**Pull request merged**